### DISCUSSION TITLE
Validation Question

### DISCUSSION THREAD

**Discussion by rmhw** [AvaloniaUI/Avalonia#13337] [Q&A]
When I add validation to a form using data annotations, the error appears when the form is first opened.

But, that isn't right - my user hasn't yet made a mistake - so I'd prefer to show validation only when the field has been changed by the user or the user has submitted the form.

Is this possible with data annotations, or would I need to use something like INotifyDataErrorInfo where an error triggers only if field is changes AND doesn't meet validation expectations.

**Comment by thevortexcloud** [comment]
Not sure about using the annotations. But if you are using ReactiveUI you can fairly easily add custom validation logic using `ValidationRule()`.

https://www.reactiveui.net/docs/handbook/user-input-validation/

**Comment by rmhw** [comment]
I think I've figured something out that works.

1) Remove the DataAnnotations validator as per the documentation
2) Add a listener to PropertyChanged in a view model base class, which triggers validation using data annotations and saves the errors in a collection.
3) Implement INotifyDataErrorInfo in a view model base class returning the results of my validation from 2), if the property has changed

Good news is, it seems to work. But am I missing a more obvious or elegant solution?

**Comment by timunie** [comment]
Remove DataAnnotationValidator and instead use the Community Toolkit validations, which basically also uses DataAnnotations but on their own. This should work better afaik.
